File: pipeline/catalog.py
#!/usr/bin/env python3
"""Build a compact, provenance-bound catalog over the complete arXiv archive."""

from __future__ import annotations

import argparse
import hashlib
import json
import math
import re
import unicodedata
from collections import Counter
from pathlib import Path

from archive import MANIFEST_NAME, read_manifest, read_shard
from catalogproof import (
    MAX_DIRECTIONS,
    MAX_SUPPORTS,
    MIN_AUTHOR_GROUPS,
    MIN_DIRECTION_SUPPORT,
    MIN_DIRECTION_YEARS,
    PUBLISHED_SUPPORTS,
    SCOPES,
    VERSION,
    catalog_hash,
    check_archive_supports,
    check_catalog,
    direction_key,
    policy_contract,
)
from files import atomic_write_text
from ontology import TOPICS, TRICKS


def parse_args() -> argparse.Namespace:
    """Parse one catalog build or validation request."""
    parser = argparse.ArgumentParser()
    parser.add_argument("--archive", type=Path, required=True)
    parser.add_argument("--output", type=Path, required=True)
    parser.add_argument("--check", action="store_true")
    parser.add_argument("--limit", type=int, default=MAX_DIRECTIONS)
    return parser.parse_args()


def file_hash(path: Path) -> str:
    """Hash one source artifact without retaining its bytes."""
    digest = hashlib.sha256()
    with path.open("rb") as stream:
        for block in iter(lambda: stream.read(1024 * 1024), b""):
            digest.update(block)
    return digest.hexdigest()


def catalog_text(value: dict) -> str:
    """Serialize stable JSON without zero-padded scientific exponents."""
    rendered = json.dumps(value, ensure_ascii=False, indent=2) + "\n"
    return re.sub(r"e([+-])0+(\d+)", r"e\1\2", rendered)


def identifiers(rows: object, ontology: dict[str, list[str]]) -> set[str]:
    """Read one already-validated archive route list defensively."""
    if not isinstance(rows, list) or not all(isinstance(row, dict) for row in rows):
        raise ValueError("Catalog routes are invalid")
    result = {row.get("id") for row in rows}
    if not all(isinstance(value, str) and value in ontology for value in result):
        raise ValueError("Catalog routes are invalid")
    return result


def author_signature(paper: dict) -> str:
    """Return a non-published stable signature for the first listed author."""
    authors = paper.get("authors")
    author = authors[0] if isinstance(authors, list) and authors else "unknown"
    value = unicodedata.normalize("NFKC", str(author)).casefold().strip()
    return hashlib.sha256(value.encode()).hexdigest()[:16]


def keep_smallest(values: set[str], value: str, limit: int) -> None:
    """Retain a deterministic bounded sample independent of shard order."""
    values.add(value)
    if len(values) > limit:
        values.remove(max(values))


def association(joint: int, subject: int, technique: int, total: int) -> float:
    """Return normalized pointwise mutual information for exact paper counts."""
    if min(joint, subject, technique, total) <= 0:
        return 0.0
    probability = joint / total
    denominator = -math.log(probability)
    if denominator == 0:
        return 1.0
    value = math.log((joint * total) / (subject * technique)) / denominator
    return round(max(-1.0, min(1.0, value)), 6)


def update_direction(
    group: dict,
    paper: dict,
    manifest_hash: str,
    shard: dict,
    row: int,
) -> None:
    """Add one distinct paper to a bounded candidate-direction aggregate."""
    group["support_count"] += 1
    year = str(paper["published"])[:4]
    group["years"].add(year)
    keep_smallest(group["authors"], author_signature(paper), MIN_AUTHOR_GROUPS)
    canonical = f"arxiv:{paper['id']}"
    rank = hashlib.sha256(f"{manifest_hash}\0{canonical}".encode()).hexdigest()
    group["supports"][rank] = {
        "id": canonical,
        "month": shard["month"],
        "path": shard["path"],
        "sha256": shard["sha256"],
        "row": row,
    }
    if len(group["supports"]) > MAX_SUPPORTS:
        del group["supports"][max(group["supports"])]


def chosen_supports(group: dict) -> list[dict]:
    """Project the bounded min-hash reservoir into exact archive references."""
    chosen = [
        group["supports"][rank]
        for rank in sorted(group["supports"])[:PUBLISHED_SUPPORTS]
    ]
    return sorted(chosen, key=lambda value: value["id"])


def build_catalog(root: Path, limit: int = MAX_DIRECTIONS) -> dict:  # noqa: C901
    """Stream every promoted shard into compact exact aggregate counts."""
    if (
        not isinstance(limit, int)
        or isinstance(limit, bool)
        or not 1 <= limit <= 10_000
    ):
        raise ValueError("Catalog direction limit must be between 1 and 10000")
    index_path = root / MANIFEST_NAME
    manifest = read_manifest(root, verify_shards=False)
    shards = manifest.get("shards", [])
    if not index_path.is_file() or not shards:
        raise ValueError("Catalog requires a promoted corpus manifest")
    manifest_hash = file_hash(index_path)
    subjects: Counter[str] = Counter()
    primary_subjects: Counter[str] = Counter()
    areas_all: Counter[str] = Counter()
    areas_scope: Counter[str] = Counter()
    techniques_all: Counter[str] = Counter()
    techniques_scope: Counter[str] = Counter()
    eligible_subjects: Counter[str] = Counter()
    eligible_techniques: Counter[str] = Counter()
    directions: dict[tuple[str, str], dict] = {}
    scanned = 0
    eligible = 0
    months: list[str] = []
    for shard in sorted(shards, key=lambda row: row.get("month", "")):
        relative = shard.get("path")
        expected = shard.get("sha256")
        month = shard.get("month")
        if not all(
            isinstance(value, str) and value for value in (relative, expected, month)
        ):
            raise ValueError("Catalog shard metadata is invalid")
        path = root / relative
        if not path.is_file() or file_hash(path) != expected:
            raise ValueError(f"Catalog shard is missing or drifted: {relative}")
        payload = read_shard(path)
        months.append(month)
        for row, paper in enumerate(payload["papers"]):
            scanned += 1
            categories = set(paper["categories"])
            topics = identifiers(paper["topics"], TOPICS)
            tricks = identifiers(paper["tricks"], TRICKS)
            subjects.update(categories)
            primary_subjects.update({paper["primary_category"]})
            areas_all.update(topics)
            techniques_all.update(tricks)
            if paper["scope"] not in SCOPES:
                continue
            eligible += 1
            areas_scope.update(topics)
            techniques_scope.update(tricks)
            eligible_subjects.update(categories)
            eligible_techniques.update(tricks)
            for subject in categories:
                for technique in tricks:
                    group = directions.setdefault(
                        (subject, technique),
                        {
                            "support_count": 0,
                            "years": set(),
                            "authors": set(),
                            "supports": {},
                        },
                    )
                    update_direction(group, paper, manifest_hash, shard, row)
    source_count = manifest.get("counts", {}).get("all")
    if scanned != source_count or months != sorted(set(months)):
        raise ValueError("Catalog coverage does not match the promoted corpus")

    candidates = []
    for (subject, technique), group in directions.items():
        if (
            group["support_count"] < MIN_DIRECTION_SUPPORT
            or len(group["years"]) < MIN_DIRECTION_YEARS
            or len(group["authors"]) < MIN_AUTHOR_GROUPS
        ):
            continue
        support_refs = chosen_supports(group)
        candidates.append(
            {
                "id": direction_key(subject, technique),
                "status": "candidate",
                "subject_id": subject,
                "technique_id": technique,
                "support_count": group["support_count"],
                "year_count": len(group["years"]),
                "independent_author_groups_at_least": MIN_AUTHOR_GROUPS,
                "npmi": association(
                    group["support_count"],
                    eligible_subjects[subject],
                    eligible_techniques[technique],
                    eligible,
                ),
                "support_ids": [support["id"] for support in support_refs],
                "support_refs": support_refs,
            }
        )
    candidates.sort(
        key=lambda row: (
            -row["support_count"],
            -row["npmi"],
            row["subject_id"],
            row["technique_id"],
        )
    )
    published = candidates[:limit]

    def family_rows(ontology: dict[str, list[str]], all_counts, scoped_counts):
        return [
            {
                "id": identifier,
                "label": identifier.replace("-", " "),
                "all_paper_count": all_counts[identifier],
                "in_scope_paper_count": scoped_counts[identifier],
            }
            for identifier in sorted(ontology)
        ]

    result = {
        "schema_version": 1,
        "generator_version": VERSION,
        "status": "corpus-derived",
        "policy": policy_contract(limit),
        "corpus": {
            "manifest_sha256": manifest_hash,
            "source_count": source_count,
            "month_count": len(shards),
        },
        "coverage": {
            "scanned_papers": scanned,
            "eligible_direction_papers": eligible,
            "scanned_months": len(months),
        },
        "counts": {
            "broad_areas": len(TOPICS),
            "technique_families": len(TRICKS),
            "arxiv_subjects": len(subjects),
            "eligible_directions": len(candidates),
            "candidate_directions": len(published),
        },
        "areas": family_rows(TOPICS, areas_all, areas_scope),
        "techniques": family_rows(TRICKS, techniques_all, techniques_scope),
        "subjects": [
            {
                "id": identifier,
                "label": identifier,
                "paper_count": subjects[identifier],
                "primary_paper_count": primary_subjects[identifier],
            }
            for identifier in sorted(subjects)
        ],
        "directions": published,
        "notice": (
            "Candidate directions are corpus-derived subject-technique crossovers, "
            "not reviewed novelty or feasibility claims."
        ),
    }
    result["content_sha256"] = catalog_hash(result)
    return check_catalog(result)


def main() -> None:
    """Build or validate one catalog artifact."""
    args = parse_args()
    if args.check:
        value = json.loads(args.output.read_text(encoding="utf-8"))
        check_catalog(value)
        manifest = read_manifest(args.archive, verify_shards=False)
        if (
            value["corpus"]["manifest_sha256"]
            != file_hash(args.archive / MANIFEST_NAME)
            or value["corpus"]["source_count"] != manifest.get("counts", {}).get("all")
            or value["corpus"]["month_count"] != len(manifest.get("shards", []))
        ):
            raise ValueError("Catalog does not describe this promoted corpus")
        check_archive_supports(value, args.archive)
        return
    value = build_catalog(args.archive, args.limit)
    atomic_write_text(args.output, catalog_text(value))


if __name__ == "__main__":
    main()

File: web/src/lib/catalog.ts
import {
  hasOnlyKeys,
  isNumber,
  isRecord,
  isString,
  isStringArray,
  type RecordValue,
} from "./guards";
import { basePath } from "./paths";

const ROOT_KEYS = new Set([
  "schema_version",
  "generator_version",
  "status",
  "content_sha256",
  "policy",
  "corpus",
  "coverage",
  "counts",
  "areas",
  "techniques",
  "subjects",
  "directions",
  "notice",
]);
const CORPUS_KEYS = new Set(["manifest_sha256", "source_count", "month_count"]);
const POLICY_KEYS = new Set([
  "digest",
  "identity_version",
  "ontology_sha256",
  "scopes",
  "min_direction_support",
  "min_direction_years",
  "min_author_groups",
  "max_directions",
  "published_supports",
]);
const COUNT_KEYS = new Set([
  "broad_areas",
  "technique_families",
  "arxiv_subjects",
  "eligible_directions",
  "candidate_directions",
]);
const LENS_KEYS = new Set(["id", "label", "all_paper_count", "in_scope_paper_count"]);
const SUBJECT_KEYS = new Set(["id", "label", "paper_count", "primary_paper_count"]);
const DIRECTION_KEYS = new Set([
  "id",
  "status",
  "subject_id",
  "technique_id",
  "support_count",
  "year_count",
  "independent_author_groups_at_least",
  "npmi",
  "support_ids",
  "support_refs",
]);
const SUPPORT_KEYS = new Set(["id", "month", "path", "sha256", "row"]);
const SUBJECT_ID = /^[A-Za-z][A-Za-z0-9.-]{1,31}$/;
const DIRECTION_ID = /^direction:[0-9a-f]{64}$/;
const SUPPORT_ID = /^arxiv:\S+$/;
const DIGEST = /^[0-9a-f]{64}$/;
const MONTH = /^[0-9]{4}-[0-9]{2}$/;
const SHARD = /^[0-9]{4}-[0-9]{2}(?:-[0-9a-f]{16})?\.json\.gz$/;

export type CatalogSummary = {
  corpusDigest: string;
  catalogDigest: string;
  policyDigest: string;
  sourceCount: number;
  broadAreas: number;
  techniqueFamilies: number;
  arxivSubjects: number;
  eligibleDirections: number;
  candidateDirections: number;
  notice: string;
};

export type CatalogLens = {
  id: string;
  label: string;
  allPaperCount: number;
  inScopePaperCount: number;
};

export type CatalogSubject = {
  id: string;
  label: string;
  paperCount: number;
  primaryPaperCount: number;
};

export type CatalogDirection = {
  id: string;
  subjectId: string;
  techniqueId: string;
  supportCount: number;
  yearCount: number;
  npmi: number;
  supportIds: string[];
};

type CatalogSupport = {
  id: string;
  month: string;
  path: string;
  sha256: string;
  row: number;
};

export type Catalog = {
  summary: CatalogSummary;
  areas: CatalogLens[];
  techniques: CatalogLens[];
  subjects: CatalogSubject[];
  directions: CatalogDirection[];
};

function whole(value: unknown): value is number {
  return isNumber(value) && Number.isInteger(value) && value >= 0;
}

function exact(value: RecordValue, keys: ReadonlySet<string>): boolean {
  return hasOnlyKeys(value, keys) && Object.keys(value).length === keys.size;
}

function readLens(value: unknown, sourceCount: number): CatalogLens | null {
  if (
    !isRecord(value) ||
    !exact(value, LENS_KEYS) ||
    !isString(value.id) ||
    !value.id ||
    !isString(value.label) ||
    !value.label ||
    !whole(value.all_paper_count) ||
    !whole(value.in_scope_paper_count) ||
    value.in_scope_paper_count > value.all_paper_count ||
    value.all_paper_count > sourceCount
  ) {
    return null;
  }
  return {
    id: value.id,
    label: value.label,
    allPaperCount: value.all_paper_count,
    inScopePaperCount: value.in_scope_paper_count,
  };
}

function readSubject(value: unknown, sourceCount: number): CatalogSubject | null {
  if (
    !isRecord(value) ||
    !exact(value, SUBJECT_KEYS) ||
    !isString(value.id) ||
    !SUBJECT_ID.test(value.id) ||
    !isString(value.label) ||
    !value.label ||
    !whole(value.paper_count) ||
    !whole(value.primary_paper_count) ||
    value.primary_paper_count > value.paper_count ||
    value.paper_count > sourceCount
  ) {
    return null;
  }
  return {
    id: value.id,
    label: value.label,
    paperCount: value.paper_count,
    primaryPaperCount: value.primary_paper_count,
  };
}

function readDirection(
  value: unknown,
  subjects: ReadonlySet<string>,
  techniques: ReadonlySet<string>,
): CatalogDirection | null {
  const supports = isRecord(value) ? value.support_refs : null;
  if (
    !isRecord(value) ||
    !exact(value, DIRECTION_KEYS) ||
    !isString(value.id) ||
    !DIRECTION_ID.test(value.id) ||
    value.status !== "candidate" ||
    !isString(value.subject_id) ||
    !subjects.has(value.subject_id) ||
    !isString(value.technique_id) ||
    !techniques.has(value.technique_id) ||
    !whole(value.support_count) ||
    value.support_count < 10 ||
    !whole(value.year_count) ||
    value.year_count < 2 ||
    value.independent_author_groups_at_least !== 3 ||
    !isNumber(value.npmi) ||
    value.npmi < -1 ||
    value.npmi > 1 ||
    !isStringArray(value.support_ids) ||
    value.support_ids.length < 1 ||
    value.support_ids.length > 6 ||
    value.support_ids.some((id) => !SUPPORT_ID.test(id)) ||
    new Set(value.support_ids).size !== value.support_ids.length ||
    !Array.isArray(supports) ||
    supports.length !== value.support_ids.length
  ) {
    return null;
  }
  const supportIds = value.support_ids;
  const references: CatalogSupport[] = [];
  for (const support of supports) {
    if (
      !isRecord(support) ||
      !exact(support, SUPPORT_KEYS) ||
      !isString(support.id) ||
      !SUPPORT_ID.test(support.id) ||
      !isString(support.month) ||
      !MONTH.test(support.month) ||
      !isString(support.path) ||
      !SHARD.test(support.path) ||
      !isString(support.sha256) ||
      !DIGEST.test(support.sha256) ||
      !whole(support.row)
    ) {
      return null;
    }
    references.push(support as CatalogSupport);
  }
  if (references.some((support, index) => support.id !== supportIds[index])) {
    return null;
  }
  return {
    id: value.id,
    subjectId: value.subject_id,
    techniqueId: value.technique_id,
    supportCount: value.support_count,
    yearCount: value.year_count,
    npmi: value.npmi,
    supportIds,
  };
}

export function readCatalogSummary(value: unknown): CatalogSummary | null {
  if (
    !isRecord(value) ||
    !hasOnlyKeys(value, ROOT_KEYS) ||
    value.schema_version !== 1 ||
    value.status !== "corpus-derived" ||
    !isRecord(value.corpus) ||
    !exact(value.corpus, CORPUS_KEYS) ||
    !isRecord(value.policy) ||
    !exact(value.policy, POLICY_KEYS) ||
    !isRecord(value.counts) ||
    !hasOnlyKeys(value.counts, COUNT_KEYS) ||
    !isString(value.notice)
  ) {
    return null;
  }
  const corpusDigest = value.corpus.manifest_sha256;
  const catalogDigest = value.content_sha256;
  const policyDigest = value.policy.digest;
  const identityVersion = value.policy.identity_version;
  const ontologyDigest = value.policy.ontology_sha256;
  const sourceCount = value.corpus.source_count;
  const broadAreas = value.counts.broad_areas;
  const techniqueFamilies = value.counts.technique_families;
  const arxivSubjects = value.counts.arxiv_subjects;
  const eligibleDirections = value.counts.eligible_directions;
  const candidateDirections = value.counts.candidate_directions;
  if (
    !isString(corpusDigest) ||
    !DIGEST.test(corpusDigest) ||
    !isString(catalogDigest) ||
    !DIGEST.test(catalogDigest) ||
    !isString(policyDigest) ||
    !DIGEST.test(policyDigest) ||
    identityVersion !== "catalog-1" ||
    !isString(ontologyDigest) ||
    !DIGEST.test(ontologyDigest) ||
    value.generator_version !== "catalog-2" ||
    !Array.isArray(value.policy.scopes) ||
    value.policy.scopes.length !== 2 ||
    value.policy.scopes[0] !== "likely" ||
    value.policy.scopes[1] !== "possible" ||
    value.policy.min_direction_support !== 10 ||
    value.policy.min_direction_years !== 2 ||
    value.policy.min_author_groups !== 3 ||
    !whole(value.policy.max_directions) ||
    value.policy.max_directions < 1 ||
    value.policy.max_directions > 10_000 ||
    value.policy.published_supports !== 6 ||
    !whole(sourceCount) ||
    !whole(broadAreas) ||
    !whole(techniqueFamilies) ||
    !whole(arxivSubjects) ||
    !whole(eligibleDirections) ||
    !whole(candidateDirections)
  ) {
    return null;
  }
  if (
    candidateDirections > eligibleDirections ||
    candidateDirections > value.policy.max_directions ||
    broadAreas < 1 ||
    techniqueFamilies < 1 ||
    arxivSubjects < 1
  ) {
    return null;
  }
  return {
    corpusDigest,
    catalogDigest,
    policyDigest,
    sourceCount,
    broadAreas,
    techniqueFamilies,
    arxivSubjects,
    eligibleDirections,
    candidateDirections,
    notice: value.notice,
  };
}

export function readCatalog(value: unknown): Catalog | null {
  const summary = readCatalogSummary(value);
  if (
    !summary ||
    !isRecord(value) ||
    !Array.isArray(value.areas) ||
    !Array.isArray(value.techniques) ||
    !Array.isArray(value.subjects) ||
    !Array.isArray(value.directions)
  ) {
    return null;
  }
  const areas = value.areas.map((row) => readLens(row, summary.sourceCount));
  const techniques = value.techniques.map((row) => readLens(row, summary.sourceCount));
  const subjects = value.subjects.map((row) => readSubject(row, summary.sourceCount));
  if (
    areas.some((row) => row === null) ||
    techniques.some((row) => row === null) ||
    subjects.some((row) => row === null)
  ) {
    return null;
  }
  const validAreas = areas as CatalogLens[];
  const validTechniques = techniques as CatalogLens[];
  const validSubjects = subjects as CatalogSubject[];
  const areaIds = new Set(validAreas.map((row) => row.id));
  const techniqueIds = new Set(validTechniques.map((row) => row.id));
  const subjectIds = new Set(validSubjects.map((row) => row.id));
  if (
    validAreas.length !== summary.broadAreas ||
    validTechniques.length !== summary.techniqueFamilies ||
    validSubjects.length !== summary.arxivSubjects ||
    areaIds.size !== validAreas.length ||
    techniqueIds.size !== validTechniques.length ||
    subjectIds.size !== validSubjects.length
  ) {
    return null;
  }
  const directions = value.directions.map((row) =>
    readDirection(row, subjectIds, techniqueIds),
  );
  if (directions.some((row) => row === null)) return null;
  const validDirections = directions as CatalogDirection[];
  const directionIds = new Set(validDirections.map((row) => row.id));
  const pairs = new Set(
    validDirections.map((row) => `${row.subjectId}\u0000${row.techniqueId}`),
  );
  if (
    validDirections.length !== summary.candidateDirections ||
    directionIds.size !== validDirections.length ||
    pairs.size !== validDirections.length
  ) {
    return null;
  }
  return {
    summary,
    areas: validAreas,
    techniques: validTechniques,
    subjects: validSubjects,
    directions: validDirections,
  };
}

async function requestCatalog(
  signal: AbortSignal | undefined,
  fetcher: typeof fetch,
  base: string,
): Promise<unknown> {
  const response = await fetcher(basePath("/data/catalog.json", base), {
    signal,
    headers: { Accept: "application/json" },
  });
  if (!response.ok) throw new Error(`Catalog request failed (${response.status})`);
  return response.json();
}

export async function fetchCatalogSummary(
  signal?: AbortSignal,
  fetcher: typeof fetch = fetch,
  base: string = import.meta.env.BASE_URL,
): Promise<CatalogSummary> {
  const summary = readCatalogSummary(await requestCatalog(signal, fetcher, base));
  if (!summary) throw new Error("Catalog response is invalid");
  return summary;
}

export async function fetchCatalog(
  signal?: AbortSignal,
  fetcher: typeof fetch = fetch,
  base: string = import.meta.env.BASE_URL,
): Promise<Catalog> {
  const catalog = readCatalog(await requestCatalog(signal, fetcher, base));
  if (!catalog) throw new Error("Catalog response is invalid");
  return catalog;
}
